data/collector.go

```diff
@@ -189,56 +384,72 @@ func (c *collector) getAndPushNextReading() {
 		return
 	}
 
-	var msg v1.SensorData
-	switch v := reading.(type) {
-	case []byte:
-		msg = v1.SensorData{
-			Metadata: &v1.SensorMetadata{
-				TimeRequested: timeRequested,
-				TimeReceived:  timeReceived,
-			},
-			Data: &v1.SensorData_Binary{
-				Binary: v,
-			},
-		}
-	default:
-		// If it's not bytes, it's a struct.
-		var pbReading *structpb.Struct
-		var err error
-
-		if reflect.TypeOf(reading) == reflect.TypeOf(pb.GetReadingsResponse{}) {
-			// We special-case the GetReadingsResponse because it already contains
-			// structpb.Values in it, and the StructToStructPb logic does not handle
-			// that cleanly.
-			topLevelMap := make(map[string]*structpb.Value)
-			topLevelMap["readings"] = structpb.NewStructValue(
-				&structpb.Struct{Fields: reading.(pb.GetReadingsResponse).Readings},
-			)
-			pbReading = &structpb.Struct{Fields: topLevelMap}
-		} else {
-			pbReading, err = protoutils.StructToStructPbIgnoreOmitEmpty(reading)
-			if err != nil {
-				c.captureErrors <- errors.Wrap(err, "error while converting reading to structpb.Struct")
-				return
-			}
-		}
+	if err := c.validateReadingType(reading.Type); err != nil {
+		c.captureErrors <- errors.Wrap(err, "capture result invalid type")
+		return
+	}
 
-		msg = v1.SensorData{
-			Metadata: &v1.SensorMetadata{
-				TimeRequested: timeRequested,
-				TimeReceived:  timeReceived,
-			},
-			Data: &v1.SensorData_Struct{
-				Struct: pbReading,
-			},
-		}
+	if err := reading.Validate(); err != nil {
+		c.captureErrors <- errors.Wrap(err, "capture result failed validation")
+		return
 	}
 
+	// var msg v1.SensorData
+	// switch v := reading.(type) {
+	// case []byte:
+	// 	msg = v1.SensorData{
+	// 		Metadata: &v1.SensorMetadata{
+	// 			TimeRequested: timeRequested,
+	// 			TimeReceived:  timeReceived,
+	// 		},
+	// 		Data: &v1.SensorData_Binary{
+	// 			Binary: v,
+	// 		},
+	// 	}
+	// default:
+	// 	// If it's not bytes, it's a struct.
+	// 	var pbReading *structpb.Struct
+	// 	var err error
+
+	// 	if reflect.TypeOf(reading) == reflect.TypeOf(pb.GetReadingsResponse{}) {
+	// 		// We special-case the GetReadingsResponse because it already contains
+	// 		// structpb.Values in it, and the StructToStructPb logic does not handle
+	// 		// that cleanly.
+	// 		topLevelMap := make(map[string]*structpb.Value)
+	// 		topLevelMap["readings"] = structpb.NewStructValue(
+	// 			&structpb.Struct{Fields: reading.(pb.GetReadingsResponse).Readings},
+	// 		)
+	// 		pbReading = &structpb.Struct{Fields: topLevelMap}
+	// 	} else {
+	// 		pbReading, err = protoutils.StructToStructPbIgnoreOmitEmpty(reading)
+	// 		if err != nil {
+	// 			c.captureErrors <- errors.Wrap(err, "error while converting reading to structpb.Struct")
+	// 			return
+	// 		}
```

Comment:
TODO: Update collectors that return GetReadings to format their response appropriately